src/filesystem/impls/filer/UrlCache.js

```diff
@@ -0,0 +1,381 @@
+/*jslint vars: true, plusplus: true, devel: true, nomen: true, regexp: true, indent: 4, maxerr: 50 */
+/*global define, URL, Blob */
+
+define(function (require, exports, module) {
+    "use strict";
+
+    var UrlParams = require("utils/UrlParams").UrlParams;
+    var StartupState = require("bramble/StartupState");
+    var _ = require("thirdparty/lodash");
+    var Async = require("utils/Async");
+    var FilerUtils = require("filesystem/impls/filer/FilerUtils");
+    var Content = require("filesystem/impls/filer/lib/content");
+    var Path = FilerUtils.Path;
+    var decodePath = FilerUtils.decodePath;
+
+    var _provider;
+
+    /**
+     * Parses URL and look for the GET parameter "cacheType"
+     * if cacheType=blob, Blob URLs are used, cacheType=cacheStorage uses Cache Storage.
+     * Not specifying a cacheType allows the default to happen.
+     */
+    function getCacheType() {
+        var params = new UrlParams();
+        params.parse();
+        return params.get("cacheType") || "";
+    }
+
+    function fixPath(path) {
+        return Path.normalize(decodePath(path));
+    }
+
+    /**
+     * UrlProviderBase manages mappings between filesystem paths and URLs.
+     */
+    function UrlProviderBase() {
+        this.paths = {};
+        this.urls = {};
+    }
+    UrlProviderBase.prototype.getUrl = function(filename) {
+        // NOTE: make sure that we always return the filename unchanged if we
+        // don't have a cached URL.  Don't return a normalized, decoded version.
+        var url = this.urls[fixPath(filename)];
+
+        // We expect this to exist, if it doesn't,
+        // return path back unchanged
+        return url || filename;
+    };
+    UrlProviderBase.prototype.getFilename = function(url) {
+        var filename = this.paths[url];
+
+        // We expect this to exist, if it doesn't,
+        // return path back unchanged
+        if(!filename) {
+            return url;
+        }
+        return filename;
+    };
+
+
+    /**
+     * BlobUrlProvider uses Blob URLs in browsers that don't support CacheStorage
+     */
+    function BlobUrlProvider(warn) {
+        UrlProviderBase.call(this);
+
+        this.baseUrl = window.location.href;
+        this.shouldRewriteUrls = true;
+
+        // Warn that some network features won't work fully
+        if(warn) {
+            console.warn("[Bramble] Your browser doesn't support Service Workers and/or CacheStorage. "  +
+                         "Some preview features may not work correctly (e.g., dynamic script loading). " +
+                         "Consider using a browser that does, see: "                                     +
+                         "https://jakearchibald.github.io/isserviceworkerready");
+        }
+
+    }
+    BlobUrlProvider.prototype = Object.create(UrlProviderBase.prototype);
+    BlobUrlProvider.prototype.constructor = BlobUrlProvider;
+
+    BlobUrlProvider.prototype.init = function(callback) {
+        _.defer(callback);
```

Comment:
We could; but it gives the two providers more similar startup characteristics, by making this a true async call.